apps/src/EditorAnnotator.js

```diff
@@ -0,0 +1,961 @@
+   *
+   * If there is no block for that line it returns `null`.
+   *
+   * @param {number} lineNumber - The one-based line number.
+   * @returns {DropletBlock|null} A reference to the block or `null` if it doesn't exist.
+   */
+  getBlockForLine(lineNumber) {
+    return null;
+  }
+
+  /**
+   * Gets a reference to every block for a given line.
+   *
+   * @param {number} lineNumber - The one-based line number.
+   * @returns {Array} A list of references to blocks, empty if none exist.
+   */
+  getBlocksForLine(lineNumber) {
+    return [];
+  }
+
+  /**
+   * Dims all the blocks in the block view.
+   *
+   * @param {DropletBlock} block The block to dim.
+   */
+  dimBlock(block) {}
+
+  /**
+   * Reverse any dimmed blocks to their original state.
+   *
+   * @param {DropletBlock} block The block to undim.
+   */
+  undimBlock(block) {}
+
+  /**
+   * Highlights the given line in the active editor.
+   *
+   * It will return some metadata that can be used in the call to
+   * `unhighlightLine` to reverse this effect.
+   *
+   * @param {number} lineNumber The line number to highlight in the active editor.
+   * @param {string} highlightClass The class name to add to the line.
+   * @returns {Object} The metadata retaining the Ace Editor marker instance.
+   */
+  highlightLine(lineNumber, highlightClass) {
+    return {};
+  }
+
+  /**
+   * Unhighlights a previously highlighted line.
+   *
+   * The incoming `info` metadata is just the information returned by the prior
+   * call to `highlightLine`.
+   *
+   * @param {Object} info The metadata about the highlighted line.
+   */
+  unhighlightLine(info) {}
+
+  /**
+   * Add a line of feedback to the editor.
+   *
+   * @param {number} lineNumber The line number (1-based index)
+   * @param {string} message The text to display next to the line.
+   * @param {string} logLevel The type of annotation ('ERROR', 'INFO', etc)
+   */
+  annotateLine(lineNumber, message, logLevel = 'INFO') {}
+
+  /**
+   * Removes annotations of a particular type.
+   *
+   * @param {string} logLevel The type of annotation to clear ('ERROR', 'INFO', etc)
+   */
+  clearAnnotations(logLevel = 'INFO') {}
+}
+
+/**
+ * This is an annotation implementation for Droplet environments.
+ *
+ * Droplet has blocks that directly represent lines of code. As such, it offers
+ * capabilities that are mostly line-based. It can retrieve references, however,
+ * to blocks based on the line number.
+ */
+export class DropletAnnotator extends Annotator {
+  static libraryName() {
+    return 'droplet';
+  }
+
+  static detect() {
+    return !!EditorAnnotator.studioApp().editor;
+  }
+
+  /**
+   * Updates the editor context with the remembered highlights and annotations.
+   */
+  refresh() {
+    this.dimBlocks();
+
+    EditorAnnotator.highlightedLines.forEach(item => {
+      const block = this.getBlockForLine(item.lineNumber);
+      if (block) {
+        this.undimBlock(block);
+      }
+    });
+  }
+
+  patch() {
```

Comment:
🤯 if we have anyone in the org who is familiar with droplet it will be great to get their review here. I know you've pulled out the code into a separate class, but since it is patching droplet it seems like the code is very much droplet territory.

I think you mentioned a convo with student learning, but just wondering since I didn't remember seeing any comments from them on the PR.

Reply:
I did, and the splitting things up was in a way to ensure that ongoing work to encapsulate these libraries wasn't going to be an issue in the broad sense. It is worrying that the patching happens, of course. The other option is to permanently patch droplet.

I did not get a strong answer on this approach. I'll ping them in [this existing thread](https://codedotorg.slack.com/archives/C03DBDN67B7/p1706142470831659).